synapse/metrics/background_process_metrics.py

```diff
@@ -225,19 +287,30 @@ async def run():
         return defer.ensureDeferred(run())
 
 
-def wrap_as_background_process(desc):
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+def wrap_as_background_process(desc: str) -> Callable[[F], F]:
     """Decorator that wraps a function that gets called as a background
     process.
 
-    Equivalent of calling the function with `run_as_background_process`
+    Equivalent to calling the function with `run_as_background_process`.
+
+    Note that the annotated return type of this function is incorrect.
+    The return type of the function, once wrapped, is actually a
+    `Deferred[Optional[T]]`.
     """
 
-    def wrap_as_background_process_inner(func):
+    # NB: Return type is incorrect and should be F with a Deferred[Optional[R]] return
+    #     We can't expression the return types correctly until Python 3.10.
```

Comment:
I'm guessing this refers to ParamSpec in https://www.python.org/dev/peps/pep-0612/ ?

If so, could we use the backport in (a sufficiently new version of) [typing_extensions](https://github.com/python/typing/blob/master/typing_extensions/README.rst) now?

Reply:
Yep! The `ParamSpec` classes are available in `typing_extensions`, but mypy doesn't actually support them yet: https://github.com/python/mypy/issues/8645

Reply:
> mypy doesn't actually support them yet

mypy rejects usage of `ParamSpec`s with `error: The first argument to Callable must be a list of types or "..."  [misc]`